dbms/src/Flash/ResourceControl/LocalAdmissionController.h

```diff
@@ -0,0 +1,173 @@
+    static constexpr int32_t HighPriorityValue = 16;
+
+    // Minus 1 because uint64 max is used as special flag.
+    static constexpr uint64_t MAX_VIRTUAL_TIME = (std::numeric_limits<uint64_t>::max() >> 4) - 1;
+
+    friend class LocalAdmissionController;
+    friend class MockLocalAdmissionController;
+
+    std::string getName() const { return name; }
+
+    void consumeResource(double ru, uint64_t cpu_time_in_ns_)
+    {
+        std::lock_guard lock(mu);
+        cpu_time_in_ns += cpu_time_in_ns_;
+        bucket->consume(ru);
+        ru_consumption_delta += ru;
+    }
+
+    // Priority greater than zero: Less number means higher priority.
+    // Zero priority means has no RU left, should not schedule this resource group at all.
+    uint64_t getPriority(uint64_t max_ru_per_sec) const
+    {
+        std::lock_guard lock(mu);
+
+        const auto remaining_token = bucket->peek();
+        if (!burstable && remaining_token <= 0.0)
+            return std::numeric_limits<uint64_t>::max();
+
+        // This should not happens because tidb will check except for unittest(test static token bucket).
+        if unlikely (user_ru_per_sec == 0)
+            return std::numeric_limits<uint64_t>::max() - 1;
+
+        double weight = static_cast<double>(max_ru_per_sec) / user_ru_per_sec;
+
+        uint64_t virtual_time = cpu_time_in_ns * weight;
+        if unlikely (virtual_time > MAX_VIRTUAL_TIME)
+            virtual_time = MAX_VIRTUAL_TIME;
+
+        uint64_t priority = (((static_cast<uint64_t>(user_priority) - 1) << 60) | virtual_time);
+
+        LOG_TRACE(
+            log,
+            "getPriority detailed info: resource group name: {}, weight: {}, virtual_time: {}, user_priority: {}, "
+            "priority: {}, remaining_token: {}",
+            name,
+            weight,
+            virtual_time,
+            user_priority,
+            priority,
+            remaining_token);
+        return priority;
+    }
+
+    void initStaticTokenBucket(uint64_t capacity = std::numeric_limits<uint64_t>::max())
+    {
+        // If token bucket is normal mode, it's static, so fill_rate is zero.
+        const double init_fill_rate = 0.0;
+        const double init_tokens = user_ru_per_sec;
+        const double init_cap = capacity;
+        bucket = std::make_unique<TokenBucket>(init_fill_rate, init_tokens, init_cap);
+        assert(
+            user_priority == LowPriorityValue || user_priority == MediumPriorityValue
+            || user_priority == HighPriorityValue);
+    }
+
+    const std::string name;
+
+    uint32_t user_priority;
+    uint64_t user_ru_per_sec;
+
+    bool burstable = false;
+
+    // Definition of the RG, e.g. RG settings, priority etc.
+    resource_manager::ResourceGroup group_pb;
+
+    mutable std::mutex mu;
+
+    // Local token bucket.
+    TokenBucketPtr bucket;
+
+    // Total used cpu_time_in_ns of this ResourceGroup.
+    uint64_t cpu_time_in_ns = 0;
+
+    double ru_consumption_delta = 0.0;
```

Comment:
Where is this used?

Reply:
For now tiflash report RU usage to prometheus directly. Later we need to report to PD, just like tidb does.